xilem/src/view/flex.rs

```diff
@@ -52,127 +53,145 @@ impl<VT, Marker> Flex<VT, Marker> {
-            cx.mark_changed();
+            ctx.mark_changed();
         }
-        let mut splice = FlexSplice { ix: 0, element };
+        // TODO: Re-use scratch space?
+        let mut splice = FlexSplice {
+            // Skip the initial spacer which is always present
+            ix: 1,
+            element,
+            scratch: AppendVec::default(),
+        };
+        self.sequence
+            .seq_rebuild(&prev.sequence, view_state, ctx, &mut splice);
+        debug_assert!(splice.scratch.into_inner().is_empty());
+    }
+
+    fn teardown(
+        &self,
+        view_state: &mut Self::ViewState,
+        ctx: &mut ViewCtx,
+        element: <Self::Element as xilem_core::ViewElement>::Mut<'_>,
+    ) {
+        let mut splice = FlexSplice {
+            // Skip the initial spacer which is always present
+            ix: 1,
+            element,
+            scratch: AppendVec::default(),
+        };
+        self.sequence.seq_teardown(view_state, ctx, &mut splice);
+        debug_assert!(splice.scratch.into_inner().is_empty());
+    }
+
+    fn message(
+        &self,
+        view_state: &mut Self::ViewState,
+        id_path: &[xilem_core::ViewId],
+        message: xilem_core::DynMessage,
+        app_state: &mut State,
+    ) -> xilem_core::MessageResult<Action> {
         self.sequence
-            .rebuild(view_state, cx, &prev.sequence, &mut splice);
+            .seq_message(view_state, id_path, message, app_state)
     }
 }
 
 struct FlexSplice<'w> {
     ix: usize,
     element: WidgetMut<'w, widget::Flex>,
+    scratch: AppendVec<Pod<Box<dyn Widget>>>,
 }
 
-impl ElementSplice for FlexSplice<'_> {
-    fn push(&mut self, element: WidgetPod<Box<dyn masonry::Widget>>) {
-        self.element.insert_child_pod(self.ix, element);
+impl ElementSplice<Pod<Box<dyn Widget>>> for FlexSplice<'_> {
+    fn push(&mut self, element: Pod<Box<dyn masonry::Widget>>) {
+        self.element.insert_child_pod(self.ix, element.inner);
         self.element.insert_default_spacer(self.ix);
         self.ix += 2;
     }
-
-    fn mutate(&mut self) -> WidgetMut<Box<dyn Widget>> {
-        #[cfg(debug_assertions)]
-        let mut iterations = 0;
-        #[cfg(debug_assertions)]
-        let max = self.element.widget.len();
-        loop {
-            #[cfg(debug_assertions)]
-            {
-                if iterations > max {
-                    panic!("Got into infinite loop in FlexSplice::mutate");
-                }
-                iterations += 1;
-            }
-            let child = self.element.child_mut(self.ix);
-            if child.is_some() {
-                break;
-            }
-            self.ix += 1;
+    fn with_scratch<R>(&mut self, f: impl FnOnce(&mut AppendVec<Pod<Box<dyn Widget>>>) -> R) -> R {
+        let ret = f(&mut self.scratch);
+        for element in self.scratch.drain() {
+            self.element.insert_child_pod(self.ix, element.inner);
+            self.element.insert_default_spacer(self.ix);
```

Comment:
Ah... That's a mistake.


```suggestion
            self.element.insert_default_spacer(self.ix + 1);
```

(Or any of the equivalent reformulations)

Reply:
Lol, fixed now, but I had to refind this bug